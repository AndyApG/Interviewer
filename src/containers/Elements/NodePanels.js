import React, { PureComponent } from 'react';
import { compose, bindActionCreators } from 'redux';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import { colorDictionary } from 'network-canvas-ui';
import { includes, map, differenceBy } from 'lodash';
import { networkNodes, makeNetworkNodesForOtherPrompts } from '../../selectors/interface';
import { getExternalData } from '../../selectors/protocol';
import { actionCreators as networkActions } from '../../ducks/modules/network';
import { makeGetPromptNodeAttributes } from '../../selectors/name-generator';
import { Panel, Panels, NodeList } from '../../components/Elements';
import { MonitorDragSource } from '../../behaviours/DragAndDrop';
import configurePanels from '../../behaviours/configurePanels';

const colorPresets = [
  colorDictionary['edge-alt-1'],
  colorDictionary['edge-alt-2'],
  colorDictionary['edge-alt-3'],
  colorDictionary['edge-alt-4'],
  colorDictionary['edge-alt-5'],
];

const getHighlight = (highlight, panelNumber) => {
  if (colorDictionary[highlight]) { return colorDictionary[highlight]; }
  if (panelNumber > 0) { return colorPresets[panelNumber % colorPresets.length]; }
  return null;
};

/**
  * Configures and renders `NodeProvider`s into panels according to the protocol config
  */
class NodePanels extends PureComponent {
  static propTypes = {
    toggleNodeAttributes: PropTypes.func.isRequired,
    removeNode: PropTypes.func.isRequired,
    activePromptAttributes: PropTypes.object.isRequired,
    newNodeAttributes: PropTypes.object.isRequired,
    isDragging: PropTypes.bool,
    externalData: PropTypes.array.isRequired,
    panels: PropTypes.array,
  };

  static defaultProps = {
    panels: [],
    isDragging: false,
  };

  onDrop = ({ meta }, dataSource) => {
    if (dataSource === 'existing') {
      this.props.toggleNodeAttributes(meta.uid, { ...this.props.activePromptAttributes });
    } else {
      this.props.removeNode(meta.uid);
    }
  }

  panelHasNodes = index => this.props.panels[index].nodes.length;

  panelIsOpen = index =>
    this.props.isDragging || this.panelHasNodes(index) > 0;

  anyPanelIsOpen = () =>
    this.props.panels
      .map((panel, index) => this.panelIsOpen(index))
      .reduce((memo, panelOpen) => memo || panelOpen, false);

  configureNodeList = ({ dataSource, originNodeIds, ...nodeListOptions }) => {
    const {
      newNodeAttributes: {
        stageId,
        promptId,
      },
    } = this.props;

    const label = node => `${node.nickname}`;

    // external, needs to check list
    // otherwise check created at details.
    const accepts = (dataSource === 'existing') ?
      ({ meta }) => (
        meta.itemType === 'EXISTING_NODE' &&
        (meta.stageId !== stageId || meta.promptId !== promptId)
      ) :
      ({ meta }) => (
        meta.itemType === 'EXISTING_NODE' &&
        includes(originNodeIds, meta.uid)
      );

    return {
      ...nodeListOptions,
      onDrop: item => this.onDrop(item, dataSource),
      itemType: 'NEW_NODE',
      accepts,
      label,
    };
  }

  renderNodePanel = (panel, index) => {
    const {
      title,
      highlight,
      ...nodeListOptions
    } = panel;

    return (
      <Panel
        title={title}
        key={index}
        highlight={getHighlight(highlight, index)}
        minimise={!this.panelIsOpen(index)}
      >
        <NodeList {...this.configureNodeList(nodeListOptions)} />
      </Panel>
    );
  }

  render() {
    return (
      <Panels minimise={!this.anyPanelIsOpen()}>
        {this.props.panels.map(this.renderNodePanel)}
      </Panels>
    );
  }
}

const getNodesForDataSource = ({ nodes, existingNodes, externalData, dataSource }) => (
  dataSource === 'existing' ?
    existingNodes :
    differenceBy(externalData[dataSource].nodes, nodes, 'uid')
);

const getOriginNodeIds = ({ existingNodes, externalData, dataSource }) => (
  dataSource === 'existing' ?
    map(existingNodes, 'uid') :
    map(externalData[dataSource].nodes, 'uid')
);

function makeMapStateToProps() {
  const getPromptNodeAttributes = makeGetPromptNodeAttributes();
  const networkNodesForOtherPrompts = makeNetworkNodesForOtherPrompts();

  return function mapStateToProps(state, props) {
    const nodes = networkNodes(state);
    const existingNodes = networkNodesForOtherPrompts(state, props);
    const externalData = getExternalData(state);

    const panels = props.panels.map(
      panel => ({
        ...panel,
        originNodeIds: getOriginNodeIds({
          existingNodes,
          externalData,
          dataSource: panel.dataSource,
        }),
        nodes: getNodesForDataSource({
          nodes,
          existingNodes,
          externalData,
          dataSource: panel.dataSource,
        }),
      }),
    );

    return {
      isDragging: state.draggable.isDragging,
      activePromptAttributes: props.prompt.additionalAttributes,
      panels,
      newNodeAttributes: getPromptNodeAttributes(state, props),
    };
  };
}

function mapDispatchToProps(dispatch) {
  return {
    toggleNodeAttributes: bindActionCreators(networkActions.toggleNodeAttributes, dispatch),
    removeNode: bindActionCreators(networkActions.removeNode, dispatch),
  };
}

export default compose(
  configurePanels,
  connect(makeMapStateToProps, mapDispatchToProps),
  MonitorDragSource(['isDragging']),
)(NodePanels);
