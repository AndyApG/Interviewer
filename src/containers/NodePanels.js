import React, { PureComponent } from 'react';
import { compose, bindActionCreators } from 'redux';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import { includes, map, differenceBy } from 'lodash';
import { getNodeLabelFunction, networkNodes, makeNetworkNodesForOtherPrompts } from '../selectors/interface';
import { getExternalData } from '../selectors/protocol';
import { actionCreators as networkActions } from '../ducks/modules/network';
import { makeGetPromptNodeAttributes } from '../selectors/name-generator';
import { Panel, Panels, NodeList } from '../components/';
import { getCSSVariableAsString } from '../utils/CSSVariables';
import { MonitorDragSource } from '../behaviours/DragAndDrop';
import configurePanels from '../behaviours/configurePanels';

const colorPresets = [
  getCSSVariableAsString('--edge-alt-1'),
  getCSSVariableAsString('--edge-alt-2'),
  getCSSVariableAsString('--edge-alt-3'),
  getCSSVariableAsString('--edge-alt-4'),
  getCSSVariableAsString('--edge-alt-5'),
];

const getHighlight = (highlight, panelNumber) => {
  if (panelNumber > 0) { return colorPresets[panelNumber % colorPresets.length]; }
  return null;
};

/**
  * Configures and renders `NodeProvider`s into panels according to the protocol config
  */
class NodePanels extends PureComponent {
  static propTypes = {
    activePromptAttributes: PropTypes.object.isRequired,
    getLabel: PropTypes.func.isRequired,
    isDragging: PropTypes.bool,
    meta: PropTypes.object,
    panels: PropTypes.array,
    prompt: PropTypes.object,
    newNodeAttributes: PropTypes.object.isRequired,
    removeNode: PropTypes.func.isRequired,
    stage: PropTypes.object,
    toggleNodeAttributes: PropTypes.func.isRequired,
  };

  static defaultProps = {
    isDragging: false,
    meta: {},
    panels: [],
    prompt: { id: null },
    stage: { id: null },
  };

  onDrop = ({ meta }, dataSource) => {
    if (dataSource === 'existing') {
      this.props.toggleNodeAttributes(meta.uid, { ...this.props.activePromptAttributes });
    } else {
      this.props.removeNode(meta.uid);
    }
  }

  isPanelEmpty = index => this.props.panels[index].nodes.length === 0;

  isPanelCompatible = index => this.props.panels[index].accepts({ meta: this.props.meta });

  isPanelOpen = index =>
    (this.props.isDragging && this.isPanelCompatible(index)) || !this.isPanelEmpty(index);

  isAnyPanelOpen = () =>
    this.props.panels
      .map((panel, index) => this.isPanelOpen(index))
      .reduce((memo, panelOpen) => memo || panelOpen, false);

  renderNodePanel = (panel, index) => {
    const stageId = this.props.stage.id;
    const promptId = this.props.prompt.id;

    const {
      title,
      highlight,
      dataSource,
      ...nodeListProps
    } = panel;

    return (
      <Panel
        title={title}
        key={index}
        highlight={getHighlight(highlight, index)}
        minimise={!this.isPanelOpen(index)}
      >
        <NodeList
          listId={`PANEL_NODE_LIST_${stageId}_${promptId}_${index}`}
          id={`PANEL_NODE_LIST_${index}`}
          {...nodeListProps}
          label={this.props.getLabel}
          itemType="NEW_NODE"
          onDrop={item => this.onDrop(item, dataSource)}
        />
      </Panel>
    );
  }

  render() {
    return (
      <Panels minimise={!this.isAnyPanelOpen()}>
        {this.props.panels.map(this.renderNodePanel)}
      </Panels>
    );
  }
}

const getNodesForDataSource = ({ nodes, existingNodes, externalData, dataSource }) => (
  dataSource === 'existing' ?
    existingNodes :
    differenceBy(externalData[dataSource].nodes, nodes, 'uid')
);

const getOriginNodeIds = ({ existingNodes, externalData, dataSource }) => (
  dataSource === 'existing' ?
    map(existingNodes, 'uid') :
    map(externalData[dataSource].nodes, 'uid')
);

function makeMapStateToProps() {
  const getPromptNodeAttributes = makeGetPromptNodeAttributes();
  const networkNodesForOtherPrompts = makeNetworkNodesForOtherPrompts();

  return function mapStateToProps(state, props) {
    const allNodes = networkNodes(state);
    const existingNodes = networkNodesForOtherPrompts(state, props);
    const externalData = getExternalData(state);
    const newNodeAttributes = getPromptNodeAttributes(state, props);

    const panels = props.panels.map((panel) => {
      const originNodeIds = getOriginNodeIds({
        existingNodes,
        externalData,
        dataSource: panel.dataSource,
      });

      const nodes = getNodesForDataSource({
        nodes: allNodes,
        existingNodes,
        externalData,
        dataSource: panel.dataSource,
      });

      const accepts = (panel.dataSource === 'existing') ?
        ({ meta }) => (
          meta.itemType === 'EXISTING_NODE' &&
          (meta.stageId !== newNodeAttributes.stageId ||
            meta.promptId !== newNodeAttributes.promptId)
        ) :
        ({ meta }) => (
          meta.itemType === 'EXISTING_NODE' &&
          includes(originNodeIds, meta.uid)
        );

      return {
        ...panel,
        originNodeIds,
        nodes,
        accepts,
      };
    });

    return {
      activePromptAttributes: props.prompt.additionalAttributes,
      getLabel: getNodeLabelFunction(state),
      newNodeAttributes,
      panels,
    };
  };
}

function mapDispatchToProps(dispatch) {
  return {
    toggleNodeAttributes: bindActionCreators(networkActions.toggleNodeAttributes, dispatch),
    removeNode: bindActionCreators(networkActions.removeNode, dispatch),
  };
}

export { NodePanels };

export default compose(
  configurePanels,
  connect(makeMapStateToProps, mapDispatchToProps),
  MonitorDragSource(['isDragging', 'meta']),
)(NodePanels);
