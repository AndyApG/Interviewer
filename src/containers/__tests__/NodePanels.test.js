/* eslint-env jest */

import React from 'react';
import { shallow } from 'enzyme';
import { NodePanels } from '../NodePanels';

jest.mock('../../ui/utils/CSSVariables');

const mockProps = {
  toggleNodeAttributes: () => {},
  removeNode: () => {},
  activePromptAttributes: {},
  newNodeAttributes: {},
  getLabel: () => 'some label',
};

describe('<NodePanels />', () => {
  it('renders ok', () => {
    const component = shallow(<NodePanels {...mockProps} />);

    expect(component).toMatchSnapshot();
  });
});
