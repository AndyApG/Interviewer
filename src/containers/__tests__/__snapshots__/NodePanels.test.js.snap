// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<NodePanels /> renders ok 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <NodePanels
    activePromptAttributes={Object {}}
    isDragging={false}
    meta={Object {}}
    newNodeAttributes={Object {}}
    panels={Array []}
    prompt={
        Object {
            "id": null,
          }
    }
    removeNode={[Function]}
    stage={
        Object {
            "id": null,
          }
    }
    toggleNodeAttributes={[Function]}
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "children": Array [],
      "minimise": true,
    },
    "ref": null,
    "rendered": Array [],
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "children": Array [],
        "minimise": true,
      },
      "ref": null,
      "rendered": Array [],
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
